core/src/main/java/org/elasticsearch/index/seqno/LocalCheckpointTracker.java

```diff
@@ -122,6 +122,19 @@ public synchronized void markSeqNoAsCompleted(final long seqNo) {
     }
 
     /**
+     * Resets the checkpoint to the specified value.
+     *
+     * @param checkpoint the local checkpoint to reset this tracker to
+     */
+    synchronized void resetCheckpoint(final long checkpoint) {
+        assert checkpoint <= this.checkpoint;
+        processedSeqNo.clear();
+        firstProcessedSeqNo = checkpoint + 1;
+        nextSeqNo = checkpoint + 1;
```

Comment:
Also, such a follow-up will introduce a test that captures the problem here, namely that if we do not do something as outlined above, in this scenario a newly promoted primary can overwrite history.